src/structured_data.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright 2015 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under (1) the MaidSafe.net Commercial License,
+// version 1.0 or later, or (2) The General Public License (GPL), version 3, depending on which
+// licence you accepted on initial access to the Software (the "Licences").
+//
+// By contributing code to the SAFE Network Software, or to this project generally, you agree to be
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.0.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+use rustc_serialize::{Encodable, Encoder, Decoder};
+use cbor;
+use error::RoutingError;
+use NameType;
+use sodiumoxide::crypto;
+use std::str;
+
+/// StructuredData
+/// These types may be stored unsigned with previous and current owner keys
+/// set to the same keys. Updates though require a signature to validate
+#[derive(Clone, RustcDecodable, RustcEncodable)]
+pub struct StructuredData {
```

Comment:
I've coded the following mock.
Point of differences:
<1> no signature parameter in `new()`. User must not be trusted to sign properly, so the function does it itself.
<2> identifier may not follow utf8 for everyone (i could give in anything and it should be possible). So changes in `name()`
<3> `data_to_sign` function should be public so that order of concatenation is centralized just in-case someone needs it to sign externally or vaults need to verify signatures with customized rule (and not `verify_previous_owner_signatures`)
<4> Previous_owner_keys is Option<...> to make it more consistent with the notion that it may not be filled.
<5> This is minor: can previous_owner_signatures be made just `signatures` ? It is sometimes confusing to have that field populated and still have previous_owner_keys contain nothing.

Other than that, the deviations in naming and return types, I'll fix at my end.

```
#[derive(Clone)]
pub struct StructuredData {
    tag_type: u64,
    identifier: ::routing::NameType,//::sodiumoxide::crypto::hash::sha512::Digest,
    version: u64,
    data: Vec<u8>,
    current_owner_keys: Vec<::sodiumoxide::crypto::sign::PublicKey>,
    previous_owner_keys: Option<Vec<::sodiumoxide::crypto::sign::PublicKey>>,
    signatures: Vec<::sodiumoxide::crypto::sign::Signature>,
}

impl StructuredData {
    pub fn new(tag_type: u64,
               identifier: ::routing::NameType,//::sodiumoxide::crypto::hash::sha512::Digest,
               version: u64,
               data: Vec<u8>,
               current_owner_keys: Vec<::sodiumoxide::crypto::sign::PublicKey>,
               previous_owner_keys: Option<Vec<::sodiumoxide::crypto::sign::PublicKey>>,
               sign_key: &::sodiumoxide::crypto::sign::SecretKey) -> StructuredData {
        let mut structured_data = StructuredData {
            tag_type: tag_type,
            identifier: identifier,
            version: version,
            data: data,
            current_owner_keys: current_owner_keys,
            previous_owner_keys: previous_owner_keys,
            signatures: vec![],
        };

        structured_data.add_signature(sign_key);
        structured_data
    }

    pub fn add_signature(&mut self, sign_key: &::sodiumoxide::crypto::sign::SecretKey) {
        let signable = self.get_signable_data();
        self.signatures.push(::sodiumoxide::crypto::sign::sign_detached(&signable[..], sign_key));
    }

    pub fn get_signable_data(&self) -> Vec<u8> {
        let mut vec = self.data.iter().chain( self.version.to_string().as_bytes().iter().chain(
                self.current_owner_keys.iter().fold(
                    Vec::<u8>::new(), |mut key_vec, key| { key_vec.extend(key.0.iter().map(|a| *a).collect::<Vec<u8>>()); key_vec }
                    ).iter())).map(|a| *a).collect::<Vec<u8>>();

        if let Some(ref previous_owner_keys) = self.previous_owner_keys {
            vec.extend(previous_owner_keys.iter().fold(Vec::<u8>::new(), |mut key_vec, key| { key_vec.extend(key.0.iter().map(|a| *a).collect::<Vec<u8>>()); key_vec }));
        }

        vec
    }

    pub fn name(&self) -> ::routing::NameType {
        use ::sodiumoxide::crypto::hash::sha512::hash;
        ::routing::NameType(
            hash(&hash(&self.identifier.0).0.iter().chain(self.tag_type.to_string().as_bytes().iter()).map(|a| *a).collect::<Vec<u8>>()[..]).0)
    }
}
```